ReSharper.FSharp/src/FSharp.Psi.Services/src/Generate/GenerateProvider.fs

```diff
@@ -324,21 +324,23 @@ type FSharpOverridingMembersBuilder() =
         use disableFormatter = new DisableCodeFormatter()
 
         let typeDecl = context.Root :?> IFSharpTypeDeclaration
-        let typeRepr = typeDecl.TypeRepresentation
 
-        match typeRepr with
+        match typeDecl.TypeRepresentation with
         | :? IUnionRepresentation as unionRepr ->
             unionRepr.UnionCasesEnumerable
             |> Seq.tryHead
             |> Option.iter EnumCaseLikeDeclarationUtil.addBarIfNeeded
-        | :? ITypeAbbreviationRepresentation as abbrRepr ->
-            if isNotNull abbrRepr.FirstChild then
-                addNodesBefore abbrRepr.FirstChild [
-                    FSharpTokenType.BAR.CreateLeafElement()
-                    Whitespace()
-                ] |> ignore
+        | :? ITypeAbbreviationRepresentation as abbrRepr when abbrRepr.CanBeUnionCase ->
+            let factory = typeDecl.CreateElementFactory()
+            let declGroup = factory.CreateModuleMember($"type U = | {abbrRepr.AbbreviatedTypeOrUnionCase.DeclaredName}")
+            let typeDeclaration = declGroup.FirstChild.As<IFSharpTypeDeclaration>()
+            let repr = typeDeclaration.TypeRepresentation.As<IUnionRepresentation>()
+            let nav = FSharpTypeDeclarationNavigator.GetByTypeRepresentation(abbrRepr)
```

Comment:
Can it ever be something other than `typeDecl`?

Reply:
You mean if `GetByTypeRepresentation()` could return another TypeDeclaration than we actually want?
As far as I understand the implementation, that shouldn't be possible.

Reply:
Then it seems this line is never needed, and you could just use `typeDecl.

Reply:
Ah, I completely missed that typeDecl has the Setter, too. Thanks!